Add SourceLink validation script - Considering only Portable PDBs

JohnTortugo opened the pull request:
Relates to #2512 

This script will run as part of async publishing for validating sourcelink information.

Iterate over all *.symbols.nupkg and then over all contained DLL, EXE and PDB files. For each file that has a Microsoft or .Net foundation copyright it will try to extract the links to the source code. If links are found the script checks that the link is accessible.

JohnTortugo commented:
This script was tested here:

https://dnceng.visualstudio.com/internal/_releaseProgress?_a=release-pipeline-progress&releaseId=4911
https://dnceng.visualstudio.com/internal/_releaseProgress?_a=release-pipeline-progress&releaseId=4909
https://dnceng.visualstudio.com/internal/_releaseProgress?_a=release-pipeline-progress&releaseId=4910

danmosemsft commented:
@anipik can review for us

jcagme commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,139 @@
+  $FailedLinks = 0
+  $SourceLinkInfos = .\sourcelink.exe print-urls $FullPath
+
+  if ($LASTEXITCODE -eq 0) {
```
>if ($LASTEXITCODE -eq 0) { [](start = 2, length = 26)

Consider using tools.ps1's functions for error management for example `ExitWithExitCode `

JohnTortugo replied:
To do that I'd have to take a dependency on tools.ps1 and since I'm planning on using this file in the publishing pipeline that means I'd need to publish tools.ps1 as well as an artifact. I'll stick with the current approach and when we make the publish pipeline restore the SDK I'll include your suggestion.

jcagme commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,139 @@
+		$Status = UrlStatusCode $Link
+
+		if ($Status -ne "200") {
+		  Write-Host "Inacessible. Return status was $Status"
```
>Return status was $Status [](start = 29, length = 25)

Nit: move this out from if/else and instead set a $statusMessage variable. Then do `Write-Host $statusMessage. Return status was $Status`

danmosemsft replied:
typo - Inaccessible

jcagme commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,139 @@
+
+function CheckSourceLinkLinks {
+  param( 
+    [string] $PackagePath		# Path to a Symbols.NuGet package
```
>		 [](start = 25, length = 2)

Nit: use the same spacing between comments

jcagme commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,139 @@
+
+  # Ensure input file exist
+  if (!(Test-Path $PackagePath)) {
+    throw "Input file does not exist: $PackagePath"
```
> throw "Input file does not exist: $PackagePath" [](start = 3, length = 48)

Probably use `ExitWithExitCode` as well

JohnTortugo replied:
Same as other message.

jcagme commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,139 @@
+  Pop-Location
+
+  if ($PassedFiles -eq 0 -and $FailedFiles -eq 0) {
+	Write-Host "| `t No file to check in this package. "
```
>file [](start = 21, length = 4)

files?

danmosemsft commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,139 @@
+	  }
+  }
+  else {
+  	Write-Host "| `t | `t No SourceLink information found."
```
Include file path?

JohnTortugo replied:
It's already included by a previous msg.

Anipik commented:
@JohnTortugo I just ran the script on a symbols package, its just running  the sourcelink on .dlls
I ran it like  
```.\sourcelink.ps1 E:\artifacts\packages\symbols  E:\temp2  C:\Users\anagniho\.dotnet\tools\```

output
```
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\lib\net46\Microsoft.Win32.Registry.dll  [from © Microsoft Corporation.  All rights reserved.]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\lib\net461\Microsoft.Win32.Registry.dll  [from © .NET Foundation and Contributors]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\lib\netstandard1.3\Microsoft.Win32.Registry.dll  [from © Microsoft Corporation.  All rights reserved.]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\lib\netstandard2.0\Microsoft.Win32.Registry.dll  [from © .NET Foundation and Contributors]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\ref\net46\Microsoft.Win32.Registry.dll  [from © Microsoft Corporation.  All rights reserved.]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\ref\net461\Microsoft.Win32.Registry.dll  [from © .NET Foundation and Contributors]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\ref\net472\Microsoft.Win32.Registry.dll  [from © .NET Foundation and Contributors]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\ref\netstandard1.3\Microsoft.Win32.Registry.dll  [from © Microsoft Corporation.  All rights reserved.]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\ref\netstandard2.0\Microsoft.Win32.Registry.dll  [from © .NET Foundation and Contributors]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\runtimes\unix\lib\netstandard2.0\Microsoft.Win32.Registry.dll  [from © .NET Foundation and Contributors]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\runtimes\win\lib\net46\Microsoft.Win32.Registry.dll  [from © Microsoft Corporation.  All rights reserved.]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\runtimes\win\lib\net461\Microsoft.Win32.Registry.dll  [from © .NET Foundation and Contributors]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\runtimes\win\lib\netstandard1.3\Microsoft.Win32.Registry.dll  [from © Microsoft Corporation.  All rights reserved.]
|        |       No SourceLink information found.
|        Checking file E:\temp2\Microsoft.Win32.Registry.4.6.0-preview5.19217.5.symbols\runtimes\win\lib\netstandard2.0\Microsoft.Win32.Registry.dll  [from © .NET Foundation and Contributors]
|        |       No SourceLink information found.
| Result: Passed.
```

am i missing something here ?

JohnTortugo commented:
> @JohnTortugo I just ran the script on a symbols package, its just running the sourcelink on .dlls

Which other files exactly did you expect it to look into? You mean PDBs and .EXE ?

Anipik commented:
yes the pdbs

JohnTortugo commented:
Where can I get these packages that you're trying on?

Anipik commented:
I send you the link on teams

Anipik commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,139 @@
+  )
+
+  $FailedLinks = 0
+  $SourceLinkInfos = .\sourcelink.exe print-urls $FullPath
```
Just curious why we are not using sourcelink test here

JohnTortugo replied:
That part of the tool is behaving weirdly, ask it for instance to test this nupkg `runtime.linux-musl-arm64.Microsoft.Private.CoreFx.NETCoreApp.4.6.0-preview5.19217.5.symbols.nupkg`. It doesn't find all PDBs in the package and the error reporting is just wrong.

The current script is better for debugging purposes and now it does http request caching. Eventually I think we'll need more customization also..

JohnTortugo commented:
Updated the script. PLTAL.

@danmosemsft @Anipik I think testing the URLs will become unfeasible. There are just too many files (about 7000) and too many links (about 74k) to check. I implemented a simple cache mechanism for checking the links and it improved about 4x, but I think it's still too slow. I'll time it and report here.

danmosemsft commented:
> @danmosemsft @Anipik I think testing the URLs will become unfeasible. There are just too many files (about 7000) and too many links (about 74k) to check. I implemented a simple cache mechanism for checking the links and it improved about 4x, but I think it's still too slow. I'll time it and report here.

@JohnTortugo are you checking them fully asynchronously? I am not a networking whiz but I would imagine that we could have 100 in flight at once?

danmosemsft commented:
We do know that once a URL is found to be valid, it is always valid - if there is any value in caching that.

If we can't check them all, I think it's fine to check just some (a few per assembly?)

JohnTortugo commented:
Actually both options crossed my mind but I thought we would end up settling with just checking whether the link was pointing to GitHub or not. 

I'll give a shot with the async approach and see how it goes. Thanks for the feedback.

danmosemsft commented:
Yes for sure I would check that all the URLs go to some github blob.

JohnTortugo commented:
Hi again folks. I've made a lot of changes to the code and now the performance is good - See numbers below. Changes made were:

- Process each package in parallel as a job (different approaches were tested, current one was the best)
- Download list of files for a specific GH commit instead of doing individual http requests
- Validation of GitHub URLs
- Ignore .resources.dll files

Anipik commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,146 @@
+  [Parameter(Mandatory=$true)][string] $ExtractPath,         # Full path to directory where the packages will be extracted during validation
+  [Parameter(Mandatory=$true)][string] $SourceLinkToolPath,  # Full path to directory where dotnet SourceLink CLI was installed
+  [Parameter(Mandatory=$true)][string] $GHRepoName,          # GitHub name of the repo including the Org. E.g., dotnet/arcade
+  [Parameter(Mandatory=$true)][string] $GHCommit             # GitHub commit SHA used to build the packages
```
Just a thought, i was wondering if we could get the some default value of commit and reponame by using one of the urls in nuget package files

JohnTortugo replied:
I thought that but I didn't come up with a good plan for choosing the package to look for the information. I don't think just selecting a random package would be a good approach. Any suggestion?

Anipik commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,146 @@
+  }
+
+  # Retrieve the list of files in the repo at that particular commit point and store them in the RepoFiles hash
+  $RepoTreeURL = -Join("https://api.github.com/repos/", $GHRepoName, "/git/trees/", $GHCommit, "?recursive=1")
```
we should add a check here for the reponame i.e it contains owner as well as repo name

JohnTortugo replied:
done

Anipik commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,146 @@
+
+  # Retrieve the list of files in the repo at that particular commit point and store them in the RepoFiles hash
+  $RepoTreeURL = -Join("https://api.github.com/repos/", $GHRepoName, "/git/trees/", $GHCommit, "?recursive=1")
+  $Data = Invoke-WebRequest $RepoTreeURL | ConvertFrom-Json | Select -ExpandProperty tree
```
we should add a try catch statement for this web request

JohnTortugo replied:
done

Anipik commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,146 @@
+  $Data = Invoke-WebRequest $RepoTreeURL | ConvertFrom-Json | Select -ExpandProperty tree
+  
+  foreach ($data in $Data) {
+    $RepoFiles[$data.path] = 1
```
We can add a check on file extensions here as well to make repoFiles set smaller, hence the search will be faster

JohnTortugo replied:
Not sure how much that will improve things but I can certainly give it a try.

JohnTortugo replied:
Included it on the current implementation

Anipik commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,146 @@
+
+        $SourceLinkInfos = .\sourcelink.exe print-urls $FullPath | Out-String
+
+        if ($LastExitError -eq 0 -and -not ([string]::IsNullOrEmpty($SourceLinkInfos))) {
```
LastExitError is not getting set to zero if the  sourceLink  command doesnot throw the error. I am not sure if the error is just local to my machine or not

JohnTortugo replied:
I used the wrong variable here. I mean to use $LASTEXITCODE

Anipik commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,146 @@
+	                
+                    # Only GitHub links are valid
+                    if ($Uri.AbsoluteURI -ne $null -and $Uri.Host -match "github") {
+                      $Status = (Invoke-WebRequest -Uri $Link -UseBasicParsing -Method HEAD -TimeoutSec 5).StatusCode
```
I will suggest to do this invoke webrequest on the original link and not after replacing it with GitHub repo and GitHub commit sha

JohnTortugo replied:
That was already the case.

Anipik commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,146 @@
+	                
+                    # Only GitHub links are valid
+                    if ($Uri.AbsoluteURI -ne $null -and $Uri.Host -match "github") {
+                      $Status = (Invoke-WebRequest -Uri $Link -UseBasicParsing -Method HEAD -TimeoutSec 5).StatusCode
```
can you please verify if this command is correct because i am getting these failed retreive errors, which should not be there ?
```
5
        File E:\MyPackage.1.0.0.symbols\content\System.Web.HttpUtility.pdb has broken links:
                Failed to retrieve https://raw.githubusercontent.com/dotnet/corefx/4ad418ef329d77bd6497d65b30579f868c185d70/src/System.Web.HttpUtility/src/System/Web/HttpUtility.cs
                Failed to retrieve https://raw.githubusercontent.com/dotnet/corefx/4ad418ef329d77bd6497d65b30579f868c185d70/src/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs
                Failed to retrieve https://raw.githubusercontent.com/dotnet/corefx/4ad418ef329d77bd6497d65b30579f868c185d70/src/System.Web.HttpUtility/src/System/Web/Util/HttpEncoderUtility.cs
                Failed to retrieve https://raw.githubusercontent.com/dotnet/corefx/4ad418ef329d77bd6497d65b30579f868c185d70/src/System.Web.HttpUtility/src/System/Web/Util/UriUtil.cs
                Failed to retrieve https://raw.githubusercontent.com/dotnet/corefx/4ad418ef329d77bd6497d65b30579f868c185d70/src/System.Web.HttpUtility/src/System/Web/Util/Utf16StringValidator.cs
The property 'value' cannot be found on this object. Verify that the property exists and can be set.
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : PropertyAssignmentException
    + PSComputerName        : localhost

1
        File E:\MyPackage.1.0.0.symbols\content\System.Xml.XPath.XDocument.pdb has broken links:
                Failed to retrieve https://raw.githubusercontent.com/dotnet/corefx/4ad418ef329d77bd6497d65b30579f868c185d70/src/System.Xml.XPath.XDocument/src/System/Xml/XPath/XDocumentExtensions.cs
The property 'value' cannot be found on this object. Verify that the property exists and can be set.
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : PropertyAssignmentException
    + PSComputerName        : localhost

Passed.

```

JohnTortugo replied:
Let me know which package you're verifying and how you're executing the script. You can send me that on Teams.

Anipik replied:
sure

danmosemsft commented on eng/common/SourceLinkValidation.ps1:
```diff
@@ -0,0 +1,166 @@
+                  }
+                }
+
+		        if ($Status -ne 200) {
```
indent

danmosemsft replied:
actually is there a way to pretty print the whole thing to get indents right..

JohnTortugo commented:
Thank you all for reviewing!
